### PR TITLE
Statistics/Achievements API addition: GameRegistry get statistics by Entity/Item/Team

### PR DESCRIPTION
**Pull request by ST-DDT**
**Helper PR for:** https://github.com/SpongePowered/SpongeAPI/pull/220
There is another PR for statistic set/change methods including events: https://github.com/SpongePowered/SpongeAPI/pull/318

Quick access methods for specific statistics.
* `getEntityStatistic(EntityStatisticType, EntityType)`
* `getItemStatistic(ItemStatisticType, ItemType)`
* `getTeamStatistic(TeamStatisticType, TextColor)`
+ default getters in the `GameRegistry`

`StatisticType` is some kind of grouping inside statistics like `"stat.craftItem"` for `"stat.craftItem.stone/wood/..."`

`StatisticUnit` is some kind of unit used to examine and display statistics like `"Walked xxx cm"`.
(as result of `"Walked " + unit.format(statsValue)`)

Based on @Aaron1011 suggestions for https://github.com/SpongePowered/SpongeAPI/pull/299
````java
gr.createCriteria(gr.getEntityStatistics(EntityStatisticTypes.HASKILLED, EntityTypes.ZOMBIE));
````
(Hope i could recreate your example properly)

[Minecraft 1.8.2](http://minecraft.gamepedia.com/1.8.2) adds many (20) new stats.

**Create/Register new Statistics**
Creating a new `Statistic` is pretty simple, although i'm not sure whether I should use a builder pattern here instead.
````java
// Create single statistic
Statistic wordsTyped = gr.registerSimpleStatistic("pluginXYZ.wordsTyped", StatisticUnits.COUNT);
// Add additional statistic to group
ItemStatistic nukesUsed = (ItemStatistic) gd.registerGroupedStatistic(ItemStatisticType.USE, "nuke", Map.of("itemType", ItemTypes.NUKE));
// Create new grouped statistic
EntityStatisticType entitiesBreadType = gd.registerStatisticType("stat.entitiesBreed", StatisticUnits.COUNT, EntityStatisticType.class);
EntityStatistic cowsBread = (EntityStatistic) gd.registerGroupedStatistic(entitiesBreadType, "cow", Map.of("entityType", EntityTypes.COW));
````

**TBD**
- `SimpleStatistic` as representation of none grouped statistics (not belonging to a `StatisticType`) as opposite of `GroupedStatistic`.
- Merge existing (simple) `Statistic`s into other/new groups (`StatisticType`s) at runtime.
  - Such as `Statistics.CAKE_SLICES_EATEN` into a newly created `ItemStatisticTypes.CONSUME`
- `AggregateStatistic`s that can be used to sum multiple `Statistic`s or get the max etc...

### CODE REVIEW

**Comment by Minecrell** [comment]
@ST-DDT Can you fix the merge conflicts of this PR?

**Comment by ST-DDT** [comment]
Done

**Comment by ST-DDT** [comment]
> [Minecraft 1.8.2](http://minecraft.gamepedia.com/1.8.2) will probably add many (20) new stats.

MC 1.8.2 has a release date in the near future (this FR).
Shall i add those 20 here now?

`ARMOR_CLEANED, BANNER_CLEANED, BEACON_INTERACTION, BREWINGSTAND_INTERACTION, CAKE_SLICES_EATEN, CAULDRON_FILLED, CAULDRON_USED, CHEST_OPENED, CRAFTING_TABLE_INTERACTION, DISPENSER_INSPECTED, DROPPER_INSPECTED, ENDERCHEST_OPENED, FLOWER_POTTED, FURNACE_INTERACTION, HOPPER_INSPECTED, ITEM_ENCHANTED, NOTEBLOCK_PLAYED, NOTEBLOCK_TUNED, RECORD_PLAYED, TRAPPED_CHEST_TRIGGERED`

````java
public static final Statistic ARMOR_CLEANED = null;
public static final Statistic BANNER_CLEANED = null;
public static final Statistic BEACON_INTERACTION = null;
public static final Statistic BREWINGSTAND_INTERACTION = null;
public static final Statistic CAKE_SLICES_EATEN = null;
public static final Statistic CAULDRON_FILLED = null;
public static final Statistic CAULDRON_USED = null;
public static final Statistic CHEST_OPENED = null;
public static final Statistic CRAFTING_TABLE_INTERACTION = null;
public static final Statistic DISPENSER_INSPECTED = null;
public static final Statistic DROPPER_INSPECTED = null;
public static final Statistic ENDERCHEST_OPENED = null;
public static final Statistic FLOWER_POTTED = null;
public static final Statistic FURNACE_INTERACTION = null;
public static final Statistic HOPPER_INSPECTED = null;
public static final Statistic ITEM_ENCHANTED = null;
public static final Statistic NOTEBLOCK_PLAYED = null;
public static final Statistic NOTEBLOCK_TUNED = null;
public static final Statistic RECORD_PLAYED = null;
public static final Statistic TRAPPED_CHEST_TRIGGERED = null;
````

**Comment by Minecrell** [comment]
@ST-DDT Yes, add them already. :)

**Comment by ST-DDT** [comment]
@Minecrell In a separate section with a comment like 
`// added in Minecraft 1.8.2`
or in alphabetical order merged with the previous ones?

**Comment by kenzierocks** [comment]
Might as well merge them.

**Comment by Minecrell** [comment]
@ST-DDT I agree with @kenzierocks, there is no point to separate them. We can just throw an exception at the end in the implementation and until we have a really stable production server we've probably already updated to 1.8.2.
Although Maybe we could use use the `@since` tag in the Javadocs of the field..

**Comment by ST-DDT** [comment]
Done, shall i also add some descriptions/javadocs to the statistics?

**Comment by kenzierocks** [comment]
That would be nice.

**Comment by ST-DDT** [comment]
I just had a look on the minecraft sources for statistics.
There is also a `MeasurementType`/`StatisticUnit`:
- `DISTANCE`
- `COUNT`
- `TIME`
- `DAMAGE` (Count * 0.1)

Shall i add that here too, or shall i create a separate PR for that?

**Comment by ST-DDT** [comment]
@kenzierocks Can you please check the javadocs in the latest commit, i'm not sure whether they are   linguistical correct.

**Comment by kenzierocks** [comment]
I think doing `The total` instead of `Total` would be much better, and "total count of times" doesn't quite read properly, although I'm not sure how to replace it.

**Comment by ST-DDT** [comment]
I started adding the API for registering new statistics. I added the methods to `GameRegistry` because the javadocs of `GameDictionary` state to much that is designed for `Item`s.
Please give me some feedback.

File: src/main/java/org/spongepowered/api/GameRegistry.java [hunk 1]
```diff
@@ -396,6 +411,185 @@
     Collection<String> getDefaultGameRules();
 
     /**
+     * Gets the {@link Statistic} with the specified id.
+     *
+     * @param id The id of the statistic to return
+     * @return The statistic with the given id or Optional.absent() if not found
+     */
+    Optional<? extends Statistic> getStatistic(String id);
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/GameRegistry.java]: Any specific reason to use `? extends Statistic` everywhere here?

Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/GameRegistry.java]: Overwriteability
That are probably remainings of my PR that uses this kind of generics everywhere in the GameRegistry.
Shall i remove it?

Comment (Minecrell) [inline src/main/java/org/spongepowered/api/GameRegistry.java]: Consistency is more important so probably yes

**Comment by ST-DDT** [comment]
Or shall i use a builder pattern here?

**Comment by ST-DDT** [comment]
Updated to fix merge problems.

**Comment by simon816** [comment]
> __TBD__
> * Merge existing `Statistic`s into other/new groups (`StatisticType`s).
>    * Such as `Statistics.CAKE_SLICES_EATEN` into a newly created `ItemStatisticTypes.CONSUMED`

While I prefer grouping then like this, to be consistent with the rest of the API, they should all be in one class.

**Comment by ST-DDT** [comment]
@simon816 Did i get you right? Do you want to merge `TeamStatisticTypes`, `EntityStatisticTypes` and `ItemStatisticTypes` in a single class `StatisticTypes`?

**Comment by simon816** [comment]
@ST-DDT Yes, that's what I was saying. ItemTypes for example isn't split into ToolItemTypes, ArmorItemTypes etc. Anyone have thoughts on whether it should be separate or together?

**Comment by ST-DDT** [comment]
Just for overview it would be 8 types in that class, probably a good solution to merge them. I will do that tomorrow if no one objects.
````java
public final StatisticTypes {

    public static final EntityStatisticType HAS_KILLED_ENTITY = null;
    public static final EntityStatisticType KILLED_BY_ENTITY = null;
    public static final ItemStatisticType CRAFT_ITEM = null;
    public static final ItemStatisticType USE_ITEM = null;
    public static final ItemStatisticType BREAK_ITEM = null;
    public static final ItemStatisticType MINE_ITEM = null;
    public static final TeamStatisticType HAS_KILLED_TEAM = null;
    public static final TeamStatisticType KILLED_BY_TEAM = null;

    private StatisticTypes() {
    }

}
````

**Comment by ST-DDT** [comment]
@simon816 The TBD you quoted stated something else than we just discussed (totally unrelated).

> - Merge existing `Statistic`s into other/new groups (`StatisticType`s).
>   - Such as `Statistics.CAKE_SLICES_EATEN` into a newly created `ItemStatisticTypes.CONSUMED`

`Statistics.CAKE_SLICES_EATEN` does not belong to any `StatisticType` by default, but plugins should be able to change that and attach it to a new `StatisticType` like `StatisticTypes.CONSUME_ITEM` at runtime.

**Comment by ST-DDT** [comment]
@simon816 Merged `StatisticTypes`

Should be ready to merge now.

**Pull request merged**